rwgame/GameConfig.cpp

```diff
@@ -62,25 +65,297 @@ std::string GameConfig::getDefaultConfigPath() {
     return ".";
 }
 
-int GameConfig::handler(void* user, const char* section, const char* name,
-                        const char* value) {
-    auto self = static_cast<GameConfig*>(user);
-#define MATCH(_s, _n) (strcmp(_s, section) == 0 && strcmp(_n, name) == 0)
-
-    if (MATCH("game", "path")) {
-        self->m_gamePath = value;
-    } else if (MATCH("game", "language")) {
-        // @todo Don't allow path seperators and relative directories
-        self->m_gameLanguage = value;
-    } else if (MATCH("input", "invert_y")) {
-        self->m_inputInvertY = atoi(value) > 0;
-    } else {
-        RW_MESSAGE("Unhandled config entry [" << section << "] " << name
-                                              << " = " << value);
-        return 0;
+std::string stripComments(const std::string &str) {
+    auto s = std::string(str, 0, str.find_first_of(";#"));
+    return s.erase(s.find_last_not_of(" \n\r\t")+1);
+}
+
+struct StringTranslator {
+    typedef std::string internal_type;
+    typedef std::string external_type;
+    boost::optional<external_type> get_value(const internal_type &str) {
+        return stripComments(str);
+    }
+    boost::optional<internal_type> put_value(const external_type &str) {
+        return str;
+    }
+};
+
+struct BoolTranslator {
+    typedef std::string internal_type;
+    typedef bool        external_type;
+    boost::optional<external_type> get_value(const internal_type &str) {
+        boost::optional<external_type> res;
+        try {
+            res = std::stoi(stripComments(str)) != 0;
+        } catch (std::invalid_argument &e) {
+        }
+        return res;
+    }
+    boost::optional<internal_type> put_value(const external_type &b) {
+        return internal_type(b ? "1" : "0");
+    }
+};
+
+struct IntTranslator {
+    typedef std::string internal_type;
+    typedef int         external_type;
+    boost::optional<external_type> get_value(const internal_type &str) {
+        boost::optional<external_type> res;
+        try {
+            res = std::stoi(stripComments(str));
+        } catch (std::invalid_argument &e) {
+        }
+        return res;
+    }
+    boost::optional<internal_type> put_value(const external_type &i) {
+        return std::to_string(i);
+    }
+};
+
+GameConfig::ParseResult GameConfig::saveConfig() {
+    auto filename = getConfigFile();
+    return parseConfig(ParseType::CONFIG, "",
+        ParseType::FILE, filename);
+}
+
+std::string GameConfig::getDefaultINIString() {
+    std::string result;
+    parseConfig(ParseType::DEFAULT, "", ParseType::STRING, result);
+    return result;
+}
+
+GameConfig::ParseResult GameConfig::parseConfig(
+    GameConfig::ParseType srcType, const std::string &source,
+    ParseType destType, std::string &destination)
+{
+    pt::ptree srcTree;
+    ParseResult parseResult(srcType, source, destType, destination);
+
+    try {
+        if (srcType == ParseType::STRING) {
+            pt::read_ini(source, srcTree);
```

Comment:
Unless I'm reading the documentation wrong, read_ini only accepts a stream or a file-path as an argument. It doesn't look like anything in this PR hits this path.